amp/shard_tracker.go

```diff
@@ -0,0 +1,161 @@
+package amp
+
+import (
+	"crypto/rand"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/record"
+	"github.com/lightningnetwork/lnd/routing/shards"
+)
+
+// Shard is an implementation of the shards.PaymentShards interface specific
+// to AMP payments.
+type Shard struct {
+	child *Child
+	mpp   *record.MPP
+	amp   *record.AMP
+}
+
+// A compile time check to ensure Shard implements the shards.PaymentShard
+// interface.
+var _ shards.PaymentShard = (*Shard)(nil)
+
+// Hash returns the hash used for the HTLC representing this AMP shard.
+func (s *Shard) Hash() lntypes.Hash {
+	return s.child.Hash
+}
+
+// MPP returns any extra MPP records that should be set for the final hop on
+// the route used by this shard.
+func (s *Shard) MPP() *record.MPP {
+	return s.mpp
+}
+
+// AMP returns any extra AMP records that should be set for the final hop on
+// the route used by this shard.
+func (s *Shard) AMP() *record.AMP {
+	return s.amp
+}
+
+// ShardTracker is an implementation of the shards.ShardTracker interface
+// that is able to generate payment shards according to the AMP splitting
+// algorithm. It can be used to generate new hashes to use for HTLCs, and also
+// cancel shares used for failed payment shards.
+type ShardTracker struct {
+	setID       [32]byte
+	paymentAddr [32]byte
+	totalAmt    lnwire.MilliSatoshi
+
+	sharer *SeedSharer
+
+	shards map[uint64]*Child
+}
+
+// A compile time check to ensure ShardTracker implements the
+// shards.ShardTracker interface.
+var _ shards.ShardTracker = (*ShardTracker)(nil)
+
+// NewShardTracker creates a new shard tracker to use for AMP payments. The
+// root shard, setID, payment address and total amount must be correctly set in
+// order for the TLV options to include with each shard to be created
+// correctly.
+func NewShardTracker(root, setID, payAddr [32]byte,
+	totalAmt lnwire.MilliSatoshi) *ShardTracker {
+
+	// Create a new seed sharer from this root.
+	rootShare := Share(root)
+	rootSharer := SeedSharerFromRoot(&rootShare)
+
+	return &ShardTracker{
+		setID:       setID,
+		paymentAddr: payAddr,
+		totalAmt:    totalAmt,
+		sharer:      rootSharer,
+		shards:      make(map[uint64]*Child),
+	}
+}
+
+var zeroShare = Share{}
+
+// NewShard registers a new attempt with the ShardTracker and returns a
+// new shard representing this attempt. This attempt's shard should be canceled
+// if it ends up not being used by the overall payment, i.e. if the attempt
+// fails.
+func (s *ShardTracker) NewShard(pid uint64, last bool) (shards.PaymentShard,
+	error) {
+
+	// If the current share is a zero share, it means we already returned
+	// the last shard.
+	if s.sharer.curr == zeroShare {
+		return nil, fmt.Errorf("splitter exhausted")
+	}
+
+	// Use a random child index.
+	var childIndex [4]byte
```

Comment:
I guess it depends on how it is being used. Currently a child gets created, and if it fails we xor the share back into the parent. This means that it will be split again most likely (which adds 32 byte randomness) before a new shard is sent.

In the scenario where it's the last shard that is being created, we can end up deriving multiple children from it. However, you must send a lot of attempts before a collision is likely. Maybe that potential privacy leak is not worth optimizing out? (since I think using the same index only would be a privacy concern)

Reply:
Yes, just wanted to bring it up. I think it is safe to assume that overall the vast majority of the payments will be single shot and in a post-hyperbitcoinization world with money streaming from every netflix subscriber - maybe collisions happen.

Reply:
And what about adding a counter field to the last share struct? That doesn't leak info about other payments to the receiver.

Reply:
at 32-bits, we expect a payment hash collision after 2^16 retries _of the same HTLC_. even at 1 sec per attempt, that's continually resending the same AMP HTLC for 18 hours. AMP HTLCs with different shares shouldn't ever collide in practice because they have plenty of entropy.

if we are really concerned about collisions, we could use a CTR mode (random offset + increment) and then we'd only get collisions on wrap around. in the prior example—after 136 years of sending the same HTLC. in the end, the worst that happens is payment hash reuse. if a new use case comes along that might require this sort of sustained retry behavior then we can easily switch to CTR without any protocol changes

Reply:
But what is wrong with my suggestion to use a counter to not have this problem at all, without sacrificing privacy?

Reply:
More, real work, for only a theoretical gain? :p

Reply:
> in a post-hyperbitcoinization world with money streaming from every netflix subscriber - maybe collisions happen.

In case it isn't clear, this index is just within the context of a given setID. So the ensure that all the world streaming netflix payment doesn't collide, we just need to ensure that the `setID` is unique, which easy given the negligible probability of colliding 256-bit integers. 

Ultimately, with the route Johan has taken here with the implementation, since we just XOR out the failed shares rather than increment the childIndex (which would result in a distinct payment hash for a same share), we don't really need to worry about this colliding in practice. The alternative would've been to get another random child index, for the same share, then try again (uses a diff payment hash, but same share). This alternative route would result in more collisions in practice.

Reply:
The context was clear. I just meant to say that if massive amounts of payments happen, the overall chance of a collision somewhere on the earth increases. It was totally clear that independent payments cannot collide. Let's say in some time frame, 2^32 independent payments happen that need a second attempt. There will then be on average one user that experiences a collision.

The out-xor'ing does not apply to the final share and I'd say that almost all AMP payments will be single-shot and have only a final share that may be retried with a different child index.

>The alternative would've been to get another random child index, for the same share, then try again (uses a diff payment hash, but same share). This alternative route would result in more collisions in practice.

So what I am proposing is to just keep a counter for each (split) share that is generated. So there will be a counter within the context of a final share. Then keep incrementing that with each attempt. Then there is no risk of collision, so in my opinion a better option.

Reply:
Is this like something you had in mind? https://github.com/lightningnetwork/lnd/pull/5255

However there is still the case that after 2^32 "last shards" the counter will wrap around and collisions will start happening.